File: frontend/src/ml/features.ts
import type { MoexCandle } from '../api/moexDirect';
import type { FeatureResult, FeatureVector } from './types';

export const MIN_CANDLES_REQUIRED = 50;

function mean(arr: number[]): number {
  if (arr.length === 0) return 0;
  return arr.reduce((s, x) => s + x, 0) / arr.length;
}

function stddev(arr: number[], m?: number): number {
  if (arr.length < 2) return 0;
  const mu = m ?? mean(arr);
  const variance = arr.reduce((s, x) => s + (x - mu) ** 2, 0) / arr.length;
  return Math.sqrt(variance);
}

function calcSma(prices: number[], period: number): number[] {
  return prices.map((_, i) =>
    i < period - 1 ? NaN : mean(prices.slice(i - period + 1, i + 1))
  );
}

function calcEma(prices: number[], period: number): number[] {
  const k = 2 / (period + 1);
  const result: number[] = [prices[0]];
  for (let i = 1; i < prices.length; i++) {
    result.push(prices[i] * k + result[i - 1] * (1 - k));
  }
  return result;
}

function safe(n: number, fallback = 0): number {
  return Number.isFinite(n) ? n : fallback;
}

export function calculateFeatures(candles: MoexCandle[]): FeatureResult {
  if (candles.length < MIN_CANDLES_REQUIRED) {
    return {
      available: false,
      reason: `Need at least ${MIN_CANDLES_REQUIRED} candles, got ${candles.length}`,
    };
  }

  // Work with last 50 candles only — no mutation of input.
  const slice = candles.slice(-MIN_CANDLES_REQUIRED);
  const n = slice.length - 1; // index of the latest candle

  const closes  = slice.map(c => c.close);
  const highs   = slice.map(c => c.high);
  const lows    = slice.map(c => c.low);
  const opens   = slice.map(c => c.open);
  const volumes = slice.map(c => c.volume);

  const close = closes[n];
  const high  = highs[n];
  const low   = lows[n];
  const open  = opens[n];

  if (!Number.isFinite(close) || close <= 0) {
    return { available: false, reason: 'Invalid latest close price' };
  }

  // ── Returns ────────────────────────────────────────────────────────────────
  const return_1  = closes[n - 1]  > 0 ? (close / closes[n - 1])  - 1 : 0;
  const return_3  = closes[n - 3]  > 0 ? (close / closes[n - 3])  - 1 : 0;
  const return_5  = closes[n - 5]  > 0 ? (close / closes[n - 5])  - 1 : 0;
  const return_10 = closes[n - 10] > 0 ? (close / closes[n - 10]) - 1 : 0;

  // ── Volatility (std of log returns) ───────────────────────────────────────
  function logRetStd(lookback: number): number {
    const logRets: number[] = [];
    for (let i = n - lookback + 1; i <= n; i++) {
      if (closes[i - 1] > 0 && closes[i] > 0) {
        logRets.push(Math.log(closes[i] / closes[i - 1]));
      }
    }
    return stddev(logRets);
  }
  const volatility_10 = logRetStd(10);
  const volatility_20 = logRetStd(20);

  // ── Candle shape ──────────────────────────────────────────────────────────
  const range = high - low;
  const body  = Math.abs(close - open);
  const candle_body_pct   = range > 0 ? body / range : 0;
  const candle_range_pct  = close > 0 ? range / close : 0;
  const upper_wick        = high - Math.max(open, close);
  const lower_wick        = Math.min(open, close) - low;
  const upper_wick_pct    = range > 0 ? upper_wick / range : 0;
  const lower_wick_pct    = range > 0 ? lower_wick / range : 0;

  // ── Volume ────────────────────────────────────────────────────────────────
  const prevVol5     = volumes.slice(n - 5, n);
  const avgVol5      = mean(prevVol5);
  const volume_change_5 = avgVol5 > 0 ? (volumes[n] / avgVol5) - 1 : 0;

  const vol20win       = volumes.slice(n - 19, n + 1);
  const meanVol20      = mean(vol20win);
  const stdVol20       = stddev(vol20win, meanVol20);
  const volume_zscore_20 = stdVol20 > 0 ? (volumes[n] - meanVol20) / stdVol20 : 0;

  // ── Moving averages ───────────────────────────────────────────────────────
  const smaArr = calcSma(closes, 20);
  const emaArr = calcEma(closes, 20);

  const sma20      = smaArr[n];
  const sma20prev  = smaArr[n - 1];
  const ema20      = emaArr[n];
  const ema20prev  = emaArr[n - 1];

  const price_vs_sma_20 = Number.isFinite(sma20) && sma20 > 0 ? (close / sma20) - 1 : 0;
  const price_vs_ema_20 = ema20 > 0 ? (close / ema20) - 1 : 0;
  const sma_20_slope    = Number.isFinite(sma20) && Number.isFinite(sma20prev) && close > 0
    ? (sma20 - sma20prev) / close
    : 0;
  const ema_20_slope    = ema20prev > 0 && close > 0 ? (ema20 - ema20prev) / close : 0;

  // ── High-low position over 20 candles ─────────────────────────────────────
  const highs20 = highs.slice(n - 19, n + 1);
  const lows20  = lows.slice(n - 19, n + 1);
  const max20   = Math.max(...highs20);
  const min20   = Math.min(...lows20);
  const high_low_position_20 = max20 > min20 ? (close - min20) / (max20 - min20) : 0.5;

  const features: FeatureVector = {
    return_1:             safe(return_1),
    return_3:             safe(return_3),
    return_5:             safe(return_5),
    return_10:            safe(return_10),
    volatility_10:        safe(volatility_10),
    volatility_20:        safe(volatility_20),
    candle_body_pct:      safe(candle_body_pct),
    candle_range_pct:     safe(candle_range_pct),
    upper_wick_pct:       safe(upper_wick_pct),
    lower_wick_pct:       safe(lower_wick_pct),
    volume_change_5:      safe(volume_change_5),
    volume_zscore_20:     safe(volume_zscore_20),
    price_vs_sma_20:      safe(price_vs_sma_20),
    price_vs_ema_20:      safe(price_vs_ema_20),
    sma_20_slope:         safe(sma_20_slope),
    ema_20_slope:         safe(ema_20_slope),
    high_low_position_20: safe(high_low_position_20, 0.5),
  };

  return { available: true, features };
}

File: ml/features.py
"""
Technical feature calculation from OHLCV candle data.

All features use only past and current candle data — no future leakage.
MACD and ATR are normalized by close price so they are scale-invariant
across tickers with different price levels.

FEATURE_COLUMNS defines the ordered list used by both build_dataset and train_catboost.
"""
import numpy as np
import pandas as pd

FEATURE_COLUMNS = [
    # Short-term returns (%)
    "return_1",
    "return_3",
    "return_5",
    "return_10",
    "return_20",
    # Rolling volatility: std of log returns (%)
    "volatility_10",
    "volatility_20",
    # Candle shape — all in % of range/price
    "candle_body_pct",
    "candle_range_pct",
    "upper_wick_pct",
    "lower_wick_pct",
    # Volume
    "volume_change_5",
    "volume_zscore_20",
    # Price vs SMA (%)
    "price_vs_sma_10",
    "price_vs_sma_20",
    "price_vs_sma_50",
    # Price vs EMA (%)
    "price_vs_ema_10",
    "price_vs_ema_20",
    "price_vs_ema_50",
    # MA slopes (% change over 5 bars)
    "sma_20_slope",
    "ema_20_slope",
    # Momentum
    "rsi_14",
    "macd_line",        # normalized by close (%)
    "macd_signal",
    "macd_hist",
    # Volatility
    "atr_14",           # normalized by close (%)
    # Bollinger Bands
    "bollinger_position_20",  # 0-1 within band
    "bollinger_width_20",     # band width / SMA (%)
    # Range position (0-1 within N-bar high/low range)
    "high_low_position_20",
    "high_low_position_50",
]


def calculate_features(df: pd.DataFrame) -> pd.DataFrame:
    """
    Add all FEATURE_COLUMNS to a copy of the input DataFrame.

    Args:
        df: DataFrame with columns [open, high, low, close, volume].
            Must be sorted ascending by datetime, deduplicated.

    Returns:
        Input DataFrame with feature columns appended. NaN rows are NOT dropped here;
        call drop_invalid_feature_rows() separately after labelling.
    """
    df = df.copy()
    c = df["close"]
    o = df["open"]
    h = df["high"]
    l = df["low"]
    v = df["volume"]

    # --- Returns (%) ---
    for n in [1, 3, 5, 10, 20]:
        df[f"return_{n}"] = c.pct_change(n) * 100

    # --- Volatility: rolling std of log returns (%) ---
    log_ret = np.log(c / c.shift(1))
    df["volatility_10"] = log_ret.rolling(10).std() * 100
    df["volatility_20"] = log_ret.rolling(20).std() * 100

    # --- Candle shape ---
    hl = h - l
    upper_body = df[["open", "close"]].max(axis=1)
    lower_body = df[["open", "close"]].min(axis=1)

    df["candle_range_pct"] = (hl / o * 100).where(o > 0)
    df["candle_body_pct"] = ((c - o).abs() / hl * 100).where(hl > 0).fillna(50.0)
    df["upper_wick_pct"] = ((h - upper_body) / hl * 100).where(hl > 0).fillna(0.0)
    df["lower_wick_pct"] = ((lower_body - l) / hl * 100).where(hl > 0).fillna(0.0)

    # --- Volume ---
    vol_mean5 = v.rolling(5).mean()
    vol_mean20 = v.rolling(20).mean()
    vol_std20 = v.rolling(20).std()
    df["volume_change_5"] = ((v / vol_mean5 - 1) * 100).where(vol_mean5 > 0)
    df["volume_zscore_20"] = ((v - vol_mean20) / vol_std20).where(vol_std20 > 0)

    # --- Price vs SMA (%) ---
    for period in [10, 20, 50]:
        sma = c.rolling(period).mean()
        df[f"price_vs_sma_{period}"] = ((c / sma - 1) * 100).where(sma > 0)

    # --- Price vs EMA (%) ---
    for period in [10, 20, 50]:
        ema = c.ewm(span=period, adjust=False).mean()
        df[f"price_vs_ema_{period}"] = ((c / ema - 1) * 100).where(ema > 0)

    # --- MA slopes: % change over previous 5 bars ---
    sma20 = c.rolling(20).mean()
    ema20 = c.ewm(span=20, adjust=False).mean()
    df["sma_20_slope"] = ((sma20 / sma20.shift(5) - 1) * 100).where(sma20.shift(5) > 0)
    df["ema_20_slope"] = ((ema20 / ema20.shift(5) - 1) * 100).where(ema20.shift(5) > 0)

    # --- RSI(14) via Wilder's EMA (com = period - 1) ---
    delta = c.diff()
    avg_gain = delta.clip(lower=0).ewm(com=13, adjust=False).mean()
    avg_loss = (-delta).clip(lower=0).ewm(com=13, adjust=False).mean()
    rs = avg_gain / avg_loss.replace(0, np.nan)
    df["rsi_14"] = 100.0 - (100.0 / (1.0 + rs))

    # --- MACD normalized by close (%) ---
    ema12 = c.ewm(span=12, adjust=False).mean()
    ema26 = c.ewm(span=26, adjust=False).mean()
    macd_norm = ((ema12 - ema26) / c * 100).where(c > 0)
    signal_norm = macd_norm.ewm(span=9, adjust=False).mean()
    df["macd_line"] = macd_norm
    df["macd_signal"] = signal_norm
    df["macd_hist"] = macd_norm - signal_norm

    # --- ATR(14) normalized by close (%) ---
    prev_close = c.shift(1)
    true_range = pd.DataFrame({
        "hl": h - l,
        "hc": (h - prev_close).abs(),
        "lc": (l - prev_close).abs(),
    }).max(axis=1)
    atr_raw = true_range.ewm(com=13, adjust=False).mean()
    df["atr_14"] = (atr_raw / c * 100).where(c > 0)

    # --- Bollinger Bands(20, 2) ---
    sma20_bb = c.rolling(20).mean()
    std20_bb = c.rolling(20).std()
    upper_bb = sma20_bb + 2 * std20_bb
    lower_bb = sma20_bb - 2 * std20_bb
    bb_range = upper_bb - lower_bb
    df["bollinger_position_20"] = ((c - lower_bb) / bb_range).where(bb_range > 0).fillna(0.5)
    df["bollinger_width_20"] = (bb_range / sma20_bb * 100).where(sma20_bb > 0)

    # --- High-Low range position (0-1) ---
    for period in [20, 50]:
        roll_high = h.rolling(period).max()
        roll_low = l.rolling(period).min()
        roll_range = roll_high - roll_low
        df[f"high_low_position_{period}"] = (
            ((c - roll_low) / roll_range).where(roll_range > 0).fillna(0.5)
        )

    return df


def drop_invalid_feature_rows(df: pd.DataFrame) -> pd.DataFrame:
    """Drop rows with NaN or inf in any FEATURE_COLUMNS column."""
    feat = df[FEATURE_COLUMNS].replace([np.inf, -np.inf], np.nan)
    bad = feat.isna().any(axis=1)
    n_dropped = bad.sum()
    if n_dropped > 0:
        print(f"  Dropping {n_dropped} rows with NaN/inf in features.")
    return df[~bad].reset_index(drop=True)
